packages/truffle-compile/index.js

```diff
@@ -280,72 +280,39 @@ function replaceLinkReferences(bytecode, linkReferences, libraryName) {
   return bytecode;
 }
 
-function orderABI(contract) {
-  var contract_definition;
-  var ordered_function_names = [];
-
-  for (var i = 0; i < contract.legacyAST.children.length; i++) {
-    var definition = contract.legacyAST.children[i];
-
-    // AST can have multiple contract definitions, make sure we have the
-    // one that matches our contract
-    if (
-      definition.name !== "ContractDefinition" ||
-      definition.attributes.name !== contract.contract_name
-    ) {
-      continue;
-    }
-
-    contract_definition = definition;
-    break;
+function orderABI({ abi, contract_name: contractName, ast }) {
+  // AST can have multiple contract definitions, make sure we have the
+  // one that matches our contract
+  const contractDefinition = ast.nodes.filter(
+    ({ nodeType, name }) =>
+      nodeType === "ContractDefinition" && name === contractName
+  )[0];
+
+  if (!contractDefinition || !contractDefinition.nodes) {
+    return abi;
   }
 
-  if (!contract_definition) return contract.abi;
-  if (!contract_definition.children) return contract.abi;
-
-  contract_definition.children.forEach(function(child) {
-    if (child.name === "FunctionDefinition") {
-      ordered_function_names.push(child.attributes.name);
-    }
-  });
+  // Find all function definitions
+  const orderedFunctionNames = contractDefinition.nodes
+    .filter(({ nodeType }) => nodeType === "FunctionDefinition")
+    .map(({ name: functionName }) => functionName);
 
   // Put function names in a hash with their order, lowest first, for speed.
-  var functions_to_remove = ordered_function_names.reduce(function(
-    obj,
-    value,
-    index
-  ) {
-    obj[value] = index;
-    return obj;
-  },
-  {});
-
-  // Filter out functions from the abi
-  var function_definitions = contract.abi.filter(function(item) {
-    return functions_to_remove[item.name] !== undefined;
-  });
-
-  // Sort removed function defintions
-  function_definitions = function_definitions.sort(function(item_a, item_b) {
-    var a = functions_to_remove[item_a.name];
-    var b = functions_to_remove[item_b.name];
-
-    if (a > b) return 1;
-    if (a < b) return -1;
-    return 0;
-  });
-
-  // Create a new ABI, placing ordered functions at the end.
-  var newABI = [];
-  contract.abi.forEach(function(item) {
-    if (functions_to_remove[item.name] !== undefined) return;
-    newABI.push(item);
-  });
-
-  // Now pop the ordered functions definitions on to the end of the abi..
-  Array.prototype.push.apply(newABI, function_definitions);
-
-  return newABI;
+  const functionIndexes = orderedFunctionNames
+    .map((functionName, index) => ({ [functionName]: index }))
+    .reduce((a, b) => Object.assign({}, a, b), {});
```

Comment:
Hahaha this is yet another way of functionally constructing objects :P

Reply:
Actually, wait, this has a real potential problem, in that it assumes function names are unique.  But Solidity allows overloading, and constructors and fallback functions both appear as `FunctionDefinition`s named `""`.  You're going to get a bunch of collisions here.  This either needs to be done a differently, or those need to be handled separately.

Reply:
Actually I guess that's ultimately OK seeing as this is only being used for sorting?  Like, now that I look, the old code had this problem as well.  I was going to request changes here but if the old code also did that I guess holding it up based on that is a bit silly.  Still, it's worth noting that really matching a function to an ABI entry is more work than just checking the name, so it's more work to really get the correct order, rather than grouping together functions with the same name.  (I have a function for exactly this sort of thing on my `wire-decoder` branch; maybe that could get moved to a separate PR so it could be used here.  Though presumably it'd need to go somewhere other than `truffle-decode-utils`.)

Reply:
Apparently this is a can of worms! I move to leave this limitation as-is for the time being, until the changes you mention are merged.